### ISSUE TITLE
reduce time complexity

### ISSUE THREAD

**Issue by IotchulindraRai** [RIDHAMKALRA01/Novanector1#1]
update code 
# Importing necessary libraries
try:
    import torch
    import torch.nn as nn
    import torch.optim as optim
    from torch.utils.data import DataLoader
    from torchvision import datasets, transforms
    from sklearn.metrics import classification_report
    from PIL import Image
except ModuleNotFoundError as e:
    raise ImportError("Required modules are missing. Please ensure PyTorch and torchvision are installed.") from e

# Step 1: Preparing the Data
# Define transformations to preprocess the images
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

# Load the CIFAR-10 dataset
train_data = datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
test_data = datasets.CIFAR10(root='./data', train=False, download=True, transform=transform)

# Create data loaders to handle batches of data
train_loader = DataLoader(train_data, batch_size=128, shuffle=True, num_workers=4, pin_memory=True)
test_loader = DataLoader(test_data, batch_size=128, shuffle=False, num_workers=4, pin_memory=True)

# Class names in CIFAR-10
class_names = ['airplane', 'automobile', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck']

# Step 2: Building the CNN Model
# Define a more optimized Convolutional Neural Network (CNN)
class OptimizedCNN(nn.Module):
    def __init__(self):
        super(OptimizedCNN, self).__init__()
        self.conv1 = nn.Conv2d(3, 64, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(64, 128, kernel_size=3, padding=1)
        self.conv3 = nn.Conv2d(128, 256, kernel_size=3, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.fc1 = nn.Linear(256 * 4 * 4, 1024)
        self.fc2 = nn.Linear(1024, 10)
        self.dropout = nn.Dropout(0.4)

    def forward(self, x):
        x = self.pool(nn.functional.relu(self.conv1(x)))
        x = self.pool(nn.functional.relu(self.conv2(x)))
        x = self.pool(nn.functional.relu(self.conv3(x)))
        x = x.view(-1, 256 * 4 * 4)
        x = nn.functional.relu(self.fc1(x))
        x = self.dropout(x)
        x = self.fc2(x)
        return x

# Create an instance of the model
model = OptimizedCNN()

# Define the loss function and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.AdamW(model.parameters(), lr=0.001, weight_decay=1e-4)

# Step 3: Training the Model
# Check if a GPU is available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)

# Train the model for 10 epochs
num_epochs = 10
for epoch in range(num_epochs):
    model.train()
    running_loss = 0.0
    for images, labels in train_loader:
        images, labels = images.to(device, non_blocking=True), labels.to(device, non_blocking=True)
        optimizer.zero_grad()
        outputs = model(images)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
    print(f"Epoch [{epoch + 1}/{num_epochs}], Loss: {running_loss / len(train_loader):.4f}")

# Step 4: Evaluating the Model
model.eval()
all_predictions = []
all_true_labels = []

with torch.no_grad():
    for images, labels in test_loader:
        images, labels = images.to(device, non_blocking=True), labels.to(device, non_blocking=True)
        outputs = model(images)
        _, predictions = torch.max(outputs, 1)
        all_predictions.extend(predictions.cpu().numpy())
        all_true_labels.extend(labels.cpu().numpy())

# Generate a classification report
print("Classification Report:")
print(classification_report(all_true_labels, all_predictions, target_names=class_names))

# Step 5: Testing on Custom Images
def preprocess_custom_image(image_path):
    img = Image.open(image_path).convert('RGB')
    img = img.resize((32, 32))
    img = transform(img).unsqueeze(0)
    return img.to(device)

# Predict the class of a custom image
def predict_custom_image(image_path):
    try:
        custom_image = preprocess_custom_image(image_path)
        model.eval()
        with torch.no_grad():
            output = model(custom_image)
            _, predicted_class = torch.max(output, 1)
        return class_names[predicted_class.item()]
    except FileNotFoundError:
        return "Error: The specified image path is invalid."

# Example usage
custom_image_path = 'I:\\AI\\cat.jpg'  # Replace with your image path
print(f"\nPredicted Class for Custom Image: {predict_custom_image(custom_image_path)}")
![Uploading Screenshot 2025-01-13 195022.png…]()
